action/action.php

```diff
@@ -107,4 +108,33 @@ public function handleCache(Doku_Event $event)
             http_status(500);
         }
     }
+
+    /**
+     * PNG download available via link created in JS (only if PNG caching is enabled)
+     *
+     * @param Doku_Event $event
+     * @return void
+     */
+    public function handlePNGDownload(Doku_Event $event)
+    {
+        if ($event->data !== 'plugin_diagrams_pngdownload') return;
+        $event->preventDefault();
+        $event->stopPropagation();
+
+        global $INPUT;
+        global $conf;
+
+        $cacheName = $INPUT->str('pngcache');
+        $downloadName = $INPUT->str('filename');
+
+        // serve cached PNG file
+        $file = $conf['cachedir'] . $cacheName;
+        if (file_exists($file)) {
+            header('Content-Type: image/png');
+            header("Content-Disposition: attachment; filename=$downloadName;");
```

Comment:
I'm not sure if this could also be a security problem, since I could possibly inject custom HTTP headers by passing a newline in $downloadName? Always better when our code sets the name.

Reply:
Hopefully fixed by using cleaned media / page ids for download names.